selene_sdk/samplers/random_positions_sampler.py

```diff
@@ -151,86 +159,214 @@ def __init__(self,
+    #  samples are taken and subsequently removed.
+
+    # If the genome is above 500 million base pairs, only a psuedoshuffle is
+    # performed on the genome to perform partitions. This is due to time
+    # constraints on shuffling larger sets.
+
+    # Samples are taken without replacement. However, if the sample size is
+    # larger than the proportion specified, there will be duplicate sample
+    # items. If this is not the intended behavior, the sample size should be
+    # chosen so as to not be larger than the proportion of the genome.
+
+    def _partition_by_proportion(self):
+        self._init_chroms()
+        if self._genome_n_bases > 500000000:
+            genome_positions_arr = self._psuedoshuffle()
+        else:
+            genome_positions_arr = np.arange(total, dtype=np.int64)
+            np.random.shuffle(genome_positions_arr)
+
+        start = 0
+        if self.test_holdout:
+            test_prop_N, validation_prop_N, training_prop_N = self._assign_proportions()
+            N_test = self._N_test
+            while N_test:
+                get_N = min(test_prop_N, N_test)
+                self._partition_ixs["test"] = genome_positions_arr[start : start + get_N]
+                N_test -= get_N
+                if N_test :
+                    start = start + get_N
+
+            start = start + get_N
+
+        else:
+            validation_prop_N, training_prop_N = _assign_proportions()
+
+        N_validation = self._N_validation
+        while N_validation:
+            get_N = min(validation_prop_N, N_validation)
+            self._partition_ixs["validate"] = genome_positions_arr[start:start + get_N]
+            N_validation -= get_N
+            if N_validation:
+                start = start + get_N
+
+
+        start = start + get_N
+        while self._N_train:
+            get_N = min(training_prop_N, self._N_train)
+            self._partition_ixs["train"] = genome_positions_arr[start:start + get_N]
+            self._N_train -= get_N
 
-    def _partition_genome_by_chromosome(self):
-        for mode in self.modes:
-            self._sample_from_mode[mode] = SampleIndices([], [])
-        for index, (chrom, len_chrom) in enumerate(self.reference_sequence.get_chr_lens()):
-            if chrom in self.validation_holdout:
-                self._sample_from_mode["validate"].indices.append(
-                    index)
-            elif self.test_holdout and chrom in self.test_holdout:
-                self._sample_from_mode["test"].indices.append(
-                    index)
-            else:
-                self._sample_from_mode["train"].indices.append(
-                    index)
 
-            self.sample_from_intervals.append(
-                (chrom,
-                 self.sequence_length,
-                 len_chrom - self.sequence_length))
-            self.interval_lengths.append(len_chrom - 2 * self.sequence_length)
+    def _partition_by_chromosome(self):
+        self._init_chroms()
```

Comment:
both partition by prop and chrom call `self._init_chroms()` so that should just happen before either of these functions are called, like in __init__ or something